Make password policy identifiers public

Captain-P-Goldfish opened the pull request:
If a password policy should be modified programmatically the constant
key identifiers to set the values should be accessible globally

stianst commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/RegisterTest.java:
```diff
@@ -187,11 +189,11 @@ public void registerPasswordPolicy() {
             @Override
             public void config(RealmManager manager, RealmModel adminstrationRealm, RealmModel appRealm) {
-                appRealm.setPasswordPolicy(new PasswordPolicy("length"));
+                appRealm.setPasswordPolicy(new PasswordPolicy(LengthPasswordPolicyProviderFactory.ID));
```
Please remove this. Tests are by intent using a string and not a constant here. When using a constant one example would be that someone by mistake renames the ID of the provider resulting in the test passing, but existing password policy strings would no longer work.

Captain-P-Goldfish replied:
done

stianst commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/RegisterTest.java:
```diff
@@ -187,11 +189,11 @@ public void registerPasswordPolicy() {
         });*/
         RealmRepresentation realm = testRealm().toRepresentation();
-        realm.setPasswordPolicy("length");
+        realm.setPasswordPolicy(LengthPasswordPolicyProviderFactory.ID);
```
Please remove this. Tests are by intent using a string and not a constant here.

stianst commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/RegisterTest.java:
```diff
@@ -460,7 +462,8 @@ public void registerUserUmlats() {
     @Test
     public void registerUserNotUsernamePasswordPolicy() {
-        adminClient.realm("test").update(RealmBuilder.create().passwordPolicy("notUsername").build());
+        adminClient.realm("test").update(RealmBuilder.create()
```
Please remove this. Tests are by intent using a string and not a constant here.

Captain-P-Goldfish replied:
done

stianst commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/RegisterTest.java:
```diff
@@ -460,7 +462,8 @@ public void registerUserUmlats() {
     public void registerUserNotUsernamePasswordPolicy() {
-        adminClient.realm("test").update(RealmBuilder.create().passwordPolicy("notUsername").build());
+        adminClient.realm("test").update(RealmBuilder.create()
+                                           .passwordPolicy(NotUsernamePasswordPolicyProviderFactory.ID).build());
```
Please remove this. Tests are by intent using a string and not a constant here.

Captain-P-Goldfish replied:
done

stianst commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetPasswordTest.java:
```diff
@@ -884,7 +885,7 @@ private void setPasswordPolicy(String policy) {
     @Test
     public void resetPasswordWithLengthPasswordPolicy() throws IOException, MessagingException {
-        setPasswordPolicy("length");
+        setPasswordPolicy(LengthPasswordPolicyProviderFactory.ID);
```
Please remove this. Tests are by intent using a string and not a constant here.

Captain-P-Goldfish replied:
done

stianst commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/policy/PasswordPolicyTest.java:
```diff
@@ -55,7 +59,7 @@ public void testLength() {
             PasswordPolicyManagerProvider policyManager = session.getProvider(PasswordPolicyManagerProvider.class);
 
-            realmModel.setPasswordPolicy(PasswordPolicy.parse(session, "length"));
+            realmModel.setPasswordPolicy(PasswordPolicy.parse(session, LengthPasswordPolicyProviderFactory.ID));
```
Please remove this. Tests are by intent using a string and not a constant here.

Captain-P-Goldfish replied:
done

stianst commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/policy/PasswordPolicyTest.java:
```diff
@@ -75,7 +79,7 @@ public void testDigits() {
             PasswordPolicyManagerProvider policyManager = session.getProvider(PasswordPolicyManagerProvider.class);
 
-            realmModel.setPasswordPolicy(PasswordPolicy.parse(session, "digits"));
+            realmModel.setPasswordPolicy(PasswordPolicy.parse(session, DigitsPasswordPolicyProviderFactory.ID));
```
Please remove this. Tests are by intent using a string and not a constant here.

Captain-P-Goldfish replied:
done

stianst commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/policy/PasswordPolicyTest.java:
```diff
@@ -166,7 +170,7 @@ public void testNotUsername() {
             PasswordPolicyManagerProvider policyManager = session.getProvider(PasswordPolicyManagerProvider.class);
 
-            realmModel.setPasswordPolicy(PasswordPolicy.parse(session, "notUsername"));
+            realmModel.setPasswordPolicy(PasswordPolicy.parse(session, NotUsernamePasswordPolicyProviderFactory.ID));
```
Please remove this. Tests are by intent using a string and not a constant here.

Captain-P-Goldfish replied:
done

stianst commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/policy/PasswordPolicyTest.java:
```diff
@@ -195,7 +199,8 @@ public void testRegexPatterns() {
             try {
-                realmModel.setPasswordPolicy(PasswordPolicy.parse(session, "regexPattern"));
+                realmModel.setPasswordPolicy(PasswordPolicy.parse(session,
+                                                                  RegexPatternsPasswordPolicyProviderFactory.ID));
```
Please remove this. Tests are by intent using a string and not a constant here.

Captain-P-Goldfish replied:
done

stianst commented on server-spi-private/src/main/java/org/keycloak/policy/DigitsPasswordPolicyProviderFactory.java:
```diff
@@ -26,7 +26,7 @@
 public class DigitsPasswordPolicyProviderFactory implements PasswordPolicyProviderFactory {
 
-    static final String ID = "digits";
+    public static final String ID = "digits";
```
Please explain why you are making this change.

Captain-P-Goldfish replied:
We currently got an assignment to create SCIM module that integrates into keycloak. And if our clients want to change the password of a user that does not fit the password policy we want to return the password policy within the error message within a json structure. 
So we want to parse the content from the password policy object within the current realm by using the corresponding constants.

I will rollback the unit tests on monday

Captain-P-Goldfish replied:
By the way I am fully aware that this is actually a violation of the purpose of the ID constants but these were the only constants wihtin the keycloak implementation that were matching the effective keys.

Captain-P-Goldfish committed:
Make password policy identifiers public

If a password policy should be modified prgorammatically the constant
key identifiers to set the values should be accessible globally
